# День первый DWthon - выйди за рамки стандартного мышления!

### Цель на сегодня: загрузка и анализ исходных данных 

Добро пожаловать на **DWthon** - **выйди за рамки стандартного мышления** 🤝  Если ты чувствуешь, что немного потерялся в быстро меняющемся мире (все вокруг куда-то движутся и ты по инерции тоже) и хочешь из череды событий найти что-то перспективное и востребованное, но не знаешь с чего начать, то у меня для Тебя хорошая новость! Твой старт может стать эффективным и даже эффектным 🥳 **Тебе повезло**, т.к. *DWthon* - это правильное и нужное место 🧠. 

Привет! Меня зовут Владимир - я основатель DataWorkshop и автор строк, которые Ты сейчас читаешь. Моя роль - помочь Тебе раскрыть свой потенциал! Мир меняется. Причем очень быстро, но если "словить волну", то можно быть на ее гребне. Покажу Тебе, что это значит на практике и шаг за шагом вместе будем решать конкретную задачу, чтобы извлечь пользу из данных (за которую другие люди готовы платить). 

Анализировать данные можно вручную. Но когда их становится очень много (особенно если речь идет про "большие данные" - анг. *big data*)- это, как правило, мало эффективно, поэтому нужно использовать другие инструменты, например, искусственный интеллект. Хотя, говоря точнее, будем обучать машину так, чтобы она нам помогала - это занятие так и называется - машинное обучение 👀.     

### Чего можешь ожидать?
*DWthon* длится (всего лишь) 5 дней и поэтому нет времени распыляться на мелочи, вдаваясь в мало практические подробности (различные хитрые формулы), в машинном обучении. Задача в том, чтобы показать Тебе, как сделать первые шаги в машинном обучении, получить конкретный результат (рабочее решение) и научить иному образу мышления. При случае, конечно же, еще вдохновить (без этого трудно двигаться дальше) и начать раскрывать Твой потенциал (а раз Твоя судьба занесла Тебя сюда, то, наверное, это не случайно!). **Когда чего-нибудь сильно захочешь, вся Вселенная будет способствовать тому, чтобы желание твое сбылось** (с) - пожалуйста, не упусти этот шанс! ☝️


### Черные дыры
Возможно, Тебе приходилось слышать про черные дыры в космосе, но я сейчас имею в виду нечто другое. Объясню. В DataWorkshop мы говорим так: "Будь осторожен и обходи стороной черные дыры!" Что это значит? В мире, где мы живем, существует множество задач, которые "съедают" Твою жизненную энергию, внимание, время, деньги и другие ресурсы, а взамен не дают ничего. Это как пылесос, который готов  бесконечно поглощать Твои ресурсы. Как часто у Тебя были такие дни, когда целый день что-то делаешь и делаешь, под конец дня чувствуешь себя очень уставшим, но в тоже время понимаешь - ничего важного сегодня не сделано. И на вопрос: "Так чем я сегодня в итоге занимался?" порой ответить не так легко (а иногда и не хочется). Помнишь такие дни и задания когда "черная дыра" ловила Тебя?

###  Наука о данных
Так вот, хочу Тебе сразу сообщить, что в машинном обучении и в науке о данных (анг. `Data Science`) - просто полно черных дыр! Здесь их так много, что нужно быть очень внимательным. На практике это значит, что очень легко быть занятым и выполнять много никому ненужной работы. Эффект нулевой (и даже отрицательный), т.к. время и другие ресурсы потрачены. И это неправильно, так не должно быть!  Именно поэтому, я, как практик, хочу заострить Твое внимание на этом.

С одной стороны я хочу научить Тебя создавать рабочие решения, используя науку о данных (в том числе машинное обучение), но в то же время предупредить, что этот мир с данными очень коварный и с радостью "засасывает" новичков в черные дыры, выжимая из них все соки! Предупреждён — значит вооружён 💪. Но предупредить мало. Вот почему был создан *DWthon*, я Тебе покажу шаг за шагом - что имею в виду и как с этим всем жить и справляться.


###  Закон Парето
DataWorkshop - это прежде всего практика. Поэтому, если слышишь от нас закон с каким-то умным названием - значит в нем есть практика. И на DWthon увидишь - как это можно применять!

Объясняя материал (машинное обучение или другое), я всегда стараюсь использовать человеческий язык, чтобы не прятаться за сложными формулировками. Но стоит заметить, что зависимость - меньшинство дает большинство и наоборот, заметили уже давно и даже этому дали свое название. Существует **закон Парето - 80/20**, который говорит о том, что 80% чего-либо (здесь Ты можешь вставить почти все, что захочешь, например, стоимость, энергия, потери и т. д.) генерируется 20% чего-либо (например, усилия, клиенты и т.д.). 

Изначально речь шла про финансовый аспект. Человек заметил, что 80% богатств принадлежит 20% людям (сейчас эта пропорция еще более нарушена).


## Низко висящие фрукты
😱 Наша задача - искать низко висящие фрукты! Зачем тянуться и неэффективно срывать "фрукты", которые находятся дальше. Гораздо лучше задействовать потенциал, который доступен здесь и сейчас.  Этим и займемся!

Будем учиться искать и высматривать👀 наиболее важные задачи, чтобы под конец DWThon получить классный результат. Такой результат, который дает конкретную ценность (выгоду) и кто-то готов за это платить.

Наконец, еще одна очень важная вещь, и будем начинать. У меня есть просьба 🙏 Не сравнивай себя с другими. На *DWthon* все - победители. Почему? Сам факт того, что Ты здесь, стараешься и пробуешь - это уже круто! Первый шаг сделан (читаешь), теперь пора сделать следующий шаг. Обращай внимание на каждый твой *DWthon* день. Что почерпнул нового - сравнивай с предыдущими днями. Измеряй, пожалуйста, только свою "дельту" развития, а не других ☝️. Так будет эффективнее! И у Тебя будет чувство удовлетворенности от быстрого продвижения вперед 🚀.


### Внимание! 
Ты учишься - и у Тебя могут быть вопросы. Это нормально! Мало того, это даже очень хорошо. Спрашивая, Ты более глубоко начинаешь лично прощупывать тему. Поэтому, помни если у Тебя возникли какие-либо трудности или сомнения, при выполнении задач, то обязательно напиши об этом в Slack (это наш "чатик" для общения). Все вопросы, которые касаются первого дня, стоит писать в комнате [dwthon_day1](https://bit.ly/2UTWe1G)

Не забудь подробно указать, в чем именно сложность. Лучше всего выложить "скриншот" (фотку своего экрана) со своим кодом, где видна проблема. Тем самым поможешь другим людям понять, что у Тебя происходит и они тогда смогут Тебе помочь!

Также старайся помогать другим. Благодаря этому, мы все вместе будем продвигаться вперед.


## Ну что... Поехали! 🛼
Сначала посмотри видео ниже.

Подсказка. Чтобы его увидеть, перемести курсор в ячейку с видео и нажми одновременно `SHIFT` +` ENTER`.

In [1]:
%%html
 <iframe style="height:500px;width:100%" src="https://www.youtube.com/embed/UkZPOYMczUQ" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>

## Видео - шаг за шагом

Чтобы еще больше Тебя поддержать, я дополнительно записал "пошаговое" видео. Советую его использовать, если Твое путешествие в IT только начинается - там будет больше комментариев по выполнению заданий. Это видео поможет "переварить" полученную информацию.

Кстати, помни о том, что мозгу человека необходимо время, чтобы (совсем) новая информация усвоилась и образовались нужные нейронные соединения. Поэтому, если что-то непонятно - это нормально, просто продолжай идти вперед. Главное идти. Точки объединятся, только позволь себе это (дай немного времени на это)!

In [ ]:
%%html
<iframe style="height:500px;width:100%" src="https://www.youtube.com/embed/PeFYE3XwqOY" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>

## Python

Начинаем кодить, т.е. программировать 👌🤩🥳.

Мы будем использовать язык программирования Python. Произносится это как "пайтон" (так звучит по английски), хотя по "старинке" в русскоязычных странах его часто называют "питон". Это достаточно простой язык, но очень полезный. На питоне можно делать очень много, и между прочим, создавать решения для машинного обучения. Кстати, по некоторым оценкам питон является языком  #1 по популярности. Хотя это не точно, все зависит от того - кто и как это считает, но однозначно можно сказать, что этот язык входит в TOP5 языков, которые полезны и дают возможность зарабатывать!


Для упрощения и ускорения разработки существует множество ранее реализованных разработок. Для того, чтобы был порядок, существует определенная иерархия: модуль, пакет, библиотеки....

- **модуль** (анг. `module`) - это файл, содержащий различные функции Python и глобальные переменные. Это просто файл с расширением .py с исполняемым кодом Python.
- **пакет** (анг. `package`) - это набор модулей. Он должен содержать файл init.py в качестве флага, чтобы интерпретатор python обрабатывал его как таковой. 
- **библиотека** (анг. `library`) - это набор пакетов.


## Библиотека

Сегодня мы познакомимся с первой библиотекой - `pandas`. Эта библиотека поможет работать с данными, для начала импортировать данные и посмотреть - что там внутри (какие столбцы, значения и т.д.).

Для того, чтобы начать работать с библиотекой, ее нужно загрузить (проимпортировать ее логику). Это делается очень просто, используя ключевое слово `import`.

In [2]:
import pandas as pd

`pd` - это псевдоним ` pandas` (делается, как правило, потому, чтобы меньше букв писать - вместо слова `pandas` будем писать всего лишь две буквы `pd`).

## Мы загружаем данные

Одной строки кода достаточно, чтобы загрузить почти миллион строк в секунду. 👍

In [3]:
df = pd.read_hdf("../input/data.h5")

df.shape

(820906, 9)

Давай проследим шаг за шагом, что здесь произошло:

1. `pd.read_hdf()` загружает наш набор данных в формате `hdf/h5` - это специальный формат данных, который используется довольно часто, но сейчас опустим детали. Прими как факт, что записать данные на Твой диск можно по-разному. Этот формат более эффективный, данные более качественно сжимаются, в результате чего меньше занимают места на Твоем диске, быстрее импортируются и еще много других полезных вещей.

2. присваиваем результат переменной `df`.

#### Что в результате?

Можно сказать по-простому, что `df`-  это обычная таблица, которая состоит из колонок и столбцов. Например, как таблички *Excel*.
В `pandas` такую табличку называют `DataFrame`. 

`df.shape` - показывает размер таблички, т.е. количество строк и столбцов в нашей (загруженной) таблице. Технически это видно, как `(820906, 9)`, что означает - таблица состоит из **820 906** строк (первое число) и **9** столбцов (второе число).

Ну что, идем дальше? 😉

## Краткая информация

Для начала познакомимся с:
- `pd.head()` - возвращает нам первые строки. По умолчанию он возвращает первые 5 строк (из нашей таблицы).
- `pd.sample()` - вернет случайные строки, по умолчанию - одна (`pd.sample (10)` вернет 10 случайных строк)

Обе функции полезны, хотя стоит признать, что просмотр случайных строк (запуск ячейки, например, 5-10 раз) может принести гораздо больше пользы. Дает шанс увидеть "что-то" интересное внутри, а не только смотреть на первые пять результатов. Иногда случайные строки называют "рандомные" или "рандомными" от английского слова `random` (что означает: случайный или произвольный).

Попробуй запустить ячейку ниже (там, где код `df.sample(10)`) 5-10 раз. Посмотри внимательно на значения, которые появляются. Попробуй "уловить" определенные зависимости. Потрать, пожалуйста, 1-3 минуты на это задание.


### 💡 Подсказка
Обычно мы запускаем ячейку `SHIFT` + `ENTER`. В результате ячейка запускается и "перескакивает" на следующую. Иногда нам просто хочется запустить и оставаться в той же самой ячейке (например, сейчас так хочется). Тогда разумнее использовать `CTRL` +` ENTER` - ячейка выполняется, но курсор остается в той же самой ячейке. Попробуй 😉.

In [9]:
df.sample(10)

order_id  customer_id  product_id  quantity  price_unit  price_total  \
518793     33534         1201        1661         6         295         1770   
377249     23803          183        3116         4         165          660   
219858     15377          345         774         6         295         1770   
325934     21138         1718        3153         2         994         1988   
581344     38115         2637        3454         1         195          195   
277946     18629          457        2964        24         145         3480   
573425     37564         1728        1659         5         295         1475   
326596     21182          219         611         1         850          850   
65725       4635          642          99         6         595         3570   
396858     24821         3340           1         1         828          828   

        country_id          order_date  is_canceled  
518793           0 2010-04-07 13:07:00        False  
377249           0 2011-11-25 17:27:00        False  
219858           0 2011-08-18 14:48:00        False  
325934           0 2011-11-03 10:30:00        False  
581344           0 2010-06-08 12:16:00        False  
277946           0 2011-10-05 12:51:00        False  
573425           0 2010-06-01 12:42:00        False  
326596           0 2011-11-03 13:50:00        False  
65725            0 2011-02-24 13:43:00        False  
396858           0 2011-12-05 17:17:00        False

😯 Возможно, Ты можешь быть удивлен последовательностью, в которой мы сначала смотрим на данные (как есть), а затем пытаемся интерпретировать наименования столбцов.

Всему свое время, нам важно эффективно двигаться вперед (помнишь про правило 80/20?). Скорее всего, если получилось запустить 5-10 раз и посмотреть на случайные строчки, то уже в голове появились идеи, что эти данные могут значить. И это правильно! 

Пойми, что в реальной жизни всегда можно просить документацию (и иногда ее даже предоставляют), но мой опыт показывает, что реальность - она другая. Документация - это своего рода план (желание), а получается как получается. Поэтому, важно как можно ближе быть к реальным фактам. Данные - они просто есть, а в головах людей есть разные теории какие данные есть (и какие могут быть значения). Верь фактам!

Ну что, попробуем немного порассуждать про столбцы и в целом о проблеме, которую будем решать?


## Интернет Магазин 

Давай подумаем (постановка такого вопроса поможет быстрее вылавливать суть): как работает любой магазин?

Есть клиенты (`customer_id`), которые покупают продукты (`product_id`) по цене за "единицу продукта" (штуки, килограммы) (`price_unit`) в количестве (`amount`), и все это объединяется как один заказ (`order_id`). Кроме того, у нас есть время, когда покупка совершалась `order_date`.

В прниципе, уже должно быть понятно, какие у нас данные, правда 🤔?

#### Пример.
Приведу пример. Сразу скажу, что числа и названия случайно выбраны. Я использую этот пример, чтобы стимулировать Твое воображение.


Маша (`customer_id=1`) делает заказ (`order_id=123`) 8 марта в 10:00:00 (`order_date=03-08-2021 10:00:00`), который состоит из следующих продуктов:
- **хлеб** (`product_id=1`) одна штука (` amount=1`) за 3 рубля (`price_unit=300`)
- **помидоры** (`product_id=2`) 5 штук (`amount=5`) всего по 2,55 рублей / шт (`price_unit=255`) **2,55* 5 = 12,75 рублей** (` price_total=1275`)
  
  
Кстати, стоит заметить, что Ты, как потребитель, имеешь право на возврат продуктов в течение некоторого времени (в Америке и Европе этим часто пользуются). В данном случае данные так записаны, что если возвращаешь даже один продукт - это означает возврат всего заказа, а не только одного продукта (это просто так кто-то выдумал, даже не трать особо времени - почему, кто-то выдумал что хочет именно так 😂). В наших данных заказы, которые были возвращены, будем обозначать как: `is_canceled=True`. В этом случае мы говорим, что заказ отменен (был возврат). Кстати, при возврате магазин полностью отдает деньги за заказ.


### ☝️ Обрати внимание, как мы сохраняем «цену» в рублях.

#### 12,75 рублей - записываем как "1275" - почему?

*делаем это по простой причине, чтобы не нужно было беспокоиться об округлении чисел. Мы пытаемся вдохновиться, как это работает на «физическом уровне» нашего мира. Самая маленькая денежная единица - «копейки» (цент и т.д.). Поэтому давай использовать это мышление и здесь, и мы сэкономим себе много времени и нервов (если хочешь, можно почитать  подробнее о проблеме - как "записывать деньги" в цифровом формате [здесь](https://bit.ly/2PI9oMt). Но помни наш подход - 80/20, т.е. мы пытаемся создавать 80% ценности при 20% усилий, поэтому давай не мучаться и пользоваться очень простым механизмом, тем самым упростим себе жизнь.*

## С чего начать анализ данных?

Как думаешь, если работаешь с бизнес-проектом, то с чего нужно начать? Ответ почти всегда один и тот же, вне зависимости от культуры и времени - «деньги» 💰. Это очень простое правило, с помощью которого Ты можешь давать ценность (помогать зарабатывать или хотя бы экономить деньги) компаниям и, благодаря этому, зарабатывать до **10 000 $** или больше в месяц. Немного об этой теме уже говорил в [видео](https://youtu.be/yUcbr0ZlEu4).

Способность приносить пользу «бизнесу» тоже интересна по другой причине. Благодаря этому работодатель будет искать именно Тебя, а не наоборот (это весьма удобно). Улови разницу! Наверное, можешь подумать, что я шучу, но, тем не менее, я говорю правду 😉. Можно даже сказать,что призываю организовать свою жизнь так, чтобы именно Ты был нужен многим компаниям (а не наоборот). Я постараюсь Тебе в этом помочь или хотя бы расскажу, как это у меня работает 🤝. Приступим к анализу данных ;)

Доход

Посчитаем, сколько денег заработал магазин. Почему это важно? Много причин, хотя бы потому, что если магазин генерирует, например, 100 рублей, то даже если мы придумаем очень блестящую систему улучшения, что с того? 

С другой стороны, если магазин генерирует 10 миллионов и более, то изменения в один процент - это уже много (100 000, не так ли?). Если делать 10 таких изменений в год, то твоя зарплата ~ $10к в месяц уже не звучит как что-то волшебное. Ты можешь увидеть причину и следствие (такой зарплаты).

Повторюсь, но давай придерживаться правила 80/20. Делать наиболее важные вещи - все остальное оставляем на потом. Нужно сосредоточиться на том, чтобы делать то, что действительно оказывает значительное влияние. Направляя свою жизненную энергию в нужное русло, мы достигаем гораздо больше.

Так, смотрим на обороты магазина.

In [10]:
df["price_total"].sum()

1816145630

`price_total` это комбинация `price_unit` * `quantity` для данного продукта.
Другими словами, нахождение суммы (функция `.sum()`) для `price_total` помогает получить информацию о доходах этого магазина. ☝️ 

Напомню, что `price_total` (а также `price_unit`) держим в «копейках». Собственно, давай сразу же добавим контекст. Магазин, который мы анализируем, находится в Великобритании и их деньги - фунты стерлингов. Один фунт это 100 пенсов (так называются их "копейки"). Другими словами `price` держим в "копейках", которые в данном случае называем "пенсы".

#### Хорошо, сколько там вышло?

Если присмотришься 👀, то увидишь, что ... **18 161 456,30**, что составляет 18+ миллионов фунтов стерлингов.🧐. Похоже, у нас есть что оптимизировать - это очень хорошо! 🥳

Теперь проверим, сколько денег ускользнуло (т.е. возвраты клиентов). В этом случае мы оставляем только строки "возвраты", применяя простой фильтр `is_canceled = True`, а затем вычисляем сумму. Смотри, как это легко и интуитивно делается в `pandas`.

In [11]:
df[ True == df.is_canceled  ]["price_total"].sum()

71973973

Оказалось, что магазину нужно было вернуть клиентам достаточно "круглую сумму" **719 739,73 фунтов стерлингов**. Но мы то радуемся и говорим - ура! Откуда такая радость? 🤦‍♂️ Потому что есть что оптимизировать или сэкономить.

Давай теперь изучим еще несколько вещей. Важные параметры:
- сколько у нас клиентов?
- сколько у нас продуктов?

В `pandas` у нас есть функция `nunique()` (☝️ обрати внимание в начале на `n`), эта функция возвращает количество уникальных значений. Посмотрим на это в действии.

In [12]:
df["customer_id"].nunique()

5879

Видим **5879**, это означает, что в столбце `customer_id` столько уникальных значений. У каждого клиента есть свой идентификатор - `customer_id`. Таким образом, уникальное количество значений в столбце `customer_id` отвечает на вопрос: "сколько в магазине клиентов?"

Понятно, правда?

## Задание 1.1

Для Тебя уже понятно, как определить количество клиентов. Теперь нужно поработать немного самостоятельно и сделать то же самое для продуктов. Мы хотим ответить на вопрос, сколько уникальных продуктов (товаров) имеется в ассортименте магазина?


### 💡 Подсказка
1. Вместо содержания`## YOUR CODE HERE` напиши свой код.
2. Скорее всего, Ты захочешь использовать столбец `product_id` 😉.
3. Вдохновись ячейкой выше, в которой мы нашли уникальное количество покупателей, используя столбец `customer_id`.

In [13]:
df["product_id"].nunique()

3878

### 💡 Дополнительная подсказка
Чтобы вставить пустую ячейку для ввода своего кода, можно воспользоваться горячими клавишами:
1. **CTRL** + **M** + **A**  - вставить ячейку сверху (Above)
2. **CTRL** + **M** + **B**  - ячейка снизу (Below)

## Задание 1.2
Поскольку мы уже знаем, сколько у нас товаров, давай проверим, сколько у нас заказов в магазине.

### 💡 Подсказка
1. Вместо `## YOUR CODE HERE` напиши свой код
2. Обрати внимание на столбец `order_id`.

In [19]:
df["order_id"].nunique()

43908

Это был план минимум. Чтобы задание было засчитано за выполненное, нужно:
1. Сделать скриншот, на котором видны решения заданий 1.1 и 1.2
2. Поместить скриншот на канал [dwthon_day1_done](https://bit.ly/3iaa6xL)
3. Получить бонус за хорошо выполненную работу 💪 и завтра будет следующая порция


### Вопросы?
Напомню. Если у Тебя возникла проблема с задачей или чего-то не понимаешь, помни, что Ты всегда можешь спросить! Чего-то не знать - это нормально и спрашивать - это хорошо :). Задавай вопросы, относящиеся ко дню DWthon1, в канале [dwthon_day1](https://bit.ly/2UTWe1G). Также, если можешь помочь другим ответить на возникшие вопросы в чате, то сделай это, пожалуйста :).

## Дополнительные задания

Эти задания необязательно делать, но когда есть желание поделать еще что-то с данными, то обязательно продолжай 👍.


### 🧠 Включим критическое мышление.
Этот навык очень необходим при анализе данных (и вообще пригодится в жизни).

Выше была задача - сколько товаров в магазине, верно? Но можем ли мы ответить на этот вопрос? Давай подумаем...

На самом деле нет! Потому что мы не знаем, сколько товаров в магазине. Что мы знаем - это какие товары продавались хотя бы раз! Это огромная разница, потому что может быть тысяча других продуктов (которых нет в наших данных / заказах), которые генерируют затраты, но не продаются вообще. Поэтому внимательно старайся "выхватывать" контекст. Важно оперировать фактами (не домыслами).

Теперь посчитаем «средний чек». Мы знаем, каковы доходы. Мы ведь тоже знаем - сколько заказов? Знаем! Тогда давай разделим одно на другое и получим информацию о "среднем чеке".

In [ ]:
df["price_total"].sum() / df["order_id"].nunique()

Мы так легко узнали, что средний чек составляет 413,63 фунта стерлингов. Довольно много, правда?

Подскажу, что в данном случае это сувенирный магазин. Возможно, вспомнишь тоже, что когда делаешь новогодние покупки-подарки для своей семьи или друзей, то, вероятно, тоже тратишься - больше чем обычно.

Хорошо, давай немного подумай самостоятельно.

## 😇 Включим творческое мышление

Какие еще 5-10 показателей (кроме "среднего чека") стоит проверить? Время на креатив, спокойно подумай и Ты обязательно что-нибудь выдумаешь интересное. Я в это верю 👍.

Тебе просто нужно найти пространство, где можно сосредоточиться и подумать, может быть, небольшая прогулка, быстрый холодный душ 🚿, медитация или что-то еще, что побудит мыслить творчески.

Например. Спроси себя так. Я ответственный за интернет магазин, какие показатели меня должны волновать больше всего и почему? Но можешь пойти дальше. Пробуй! Экспериментируй, в этом смысл данного задания - находить менее очевидные на первый взгляд вещи (хотя когда уже нащупаешь их, то они кажутся очень логичными).

## 🤝🗣️ Совместное обсуждение и обмен опытом

Поделись, пожалуйста, своими мыслями с другими. Зачем? Благодаря этому могут родиться еще более качественные (совместные) идеи. Это когда 1+1 = 11 😉

Напиши в чатике [dwthon_day1_ideas](https://bit.ly/3zIbBsN), что у Тебя получилось выдумать. Поучаствуй в обсуждении, и увидишь на примере - как легко генерировать множество идей в сотрудничестве с другими. Это очень полезный навык, который скорее всего Тебе пригодится в современном мире.

## Отзыв и дельта Твоего прогресса

Дай мне, пожалуйста, знать, что Тебе понравилось, было ли трудно (и где именно) и зафиксируй также для себя - чему получилось научиться на этом уроке.

### Тут 👉👉 [анкета](https://bit.ly/3j0lLyq) 👈👈
Твой отзыв позволит улучшить материал!

## До завтра! 🙌